### PR TITLE
SDI-527 fix null pointers

### PR DESCRIPTION
**Pull request by mikehalmamoj**
We think that using a multi-threaded queue listener with a non-thread safe local event store was corrupting the event store thus causing null messages. To fix this we will oknly handle one message at a time.

Also keep Redis in sync with the local event store so the same messages are loaded after a restart.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt [hunk 1]
```diff
@@ -16,7 +17,8 @@ import uk.gov.justice.hmpps.offenderevents.service.OffenderEventStore.Topics.UNK
 @Service
 class OffenderEventStore(
   @Value("\${model.cacheSize}") private val cacheSize: Int,
-  val store: MutableList<DisplayMessage> = mutableListOf()
+  val eventRepository: EventRepository,
+  val store: ArrayDeque<DisplayMessage> = ArrayDeque(),
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt:21]: Change the local event store to a Deque to make it obvious what's happening. I considered using Java's ConcurrentLinkedDeque which should have handled the multi-threaded queue listener, but opted for simplicity instead.

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt [hunk 2]
```diff
@@ -35,8 +37,13 @@ class OffenderEventStore(
     val log = LoggerFactory.getLogger(this::class.java)
   }
 
-  fun add(element: DisplayMessage) = store.add(element)
-    .also { if (store.size > cacheSize) store.removeAt(0) }
+  fun add(element: DisplayMessage) = store.addLast(element)
+    .also {
+      if (store.size > cacheSize) {
+        val removed = store.removeFirst()
+        eventRepository.deleteById(removed.messageId)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventStore.kt:44]: Also remove the message from Redis so it is synchronised with the local store.

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/resource/MessagesController.kt [hunk 3]
```diff
@@ -8,7 +8,7 @@ import org.springframework.web.bind.annotation.RequestParam
 import org.springframework.web.servlet.ModelAndView
 import uk.gov.justice.hmpps.offenderevents.service.OffenderEventStore
 
-data class DisplayMessage(val eventType: String, val messageDetails: Map<String, String?> = mapOf(), val topic: String)
+data class DisplayMessage(val eventType: String, val messageId: String, val messageDetails: Map<String, String?> = mapOf(), val topic: String)
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/resource/MessagesController.kt:11]: Also need the message Id so we can link back to the Redis cache (which is keyed by messageId).

File: src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventListener.kt [hunk 4]
```diff
@@ -25,7 +25,7 @@ class OffenderEventListener(
     val log: Logger = LoggerFactory.getLogger(this::class.java)
   }
 
-  @SqsListener(queueNames = ["event"], factory = "hmppsQueueContainerFactoryProxy")
+  @SqsListener(queueNames = ["event"], factory = "hmppsQueueContainerFactoryProxy", maxInflightMessagesPerQueue = "1", maxMessagesPerPoll = "1")
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventListener.kt:28]: Ignore async capabilities by only allowing 1 message at a time.

Comment (petergphillips) [reply src/main/kotlin/uk/gov/justice/hmpps/offenderevents/service/OffenderEventListener.kt:28]: don't think I would call it `async`. more multi threaded.

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**